### PR TITLE
Remove test skips for 117165

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123469
* #123619
* __->__ #123406
* #124309
* #123405
* #123404
* #123324

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -1548,7 +1548,11 @@ def getPublicAttrs(obj):
 
             # Make some state
             for _ in range(3):
-                optimizer.step(closure)
+                if torch._dynamo.is_compiling() and not optim_info.step_requires_closure:
+                    closure()
+                    optimizer.step()
+                else:
+                    optimizer.step(closure)
```
Comment (janeyx99) [inline test/test_optim.py:1555]: This is more readable. Would be good to have this as the pattern everywhere.
```suggestion
                if optim_info.step_requires_closure:
                    optimizer.step(closure)
                else:
                    closure()
                    optimizer.step()
```

**Review by janeyx99 — COMMENTED** [review]
Did not give a thorough review, will do after CI is rectified

File: torch/optim/adadelta.py [hunk 2]
```diff
@@ -43,6 +43,8 @@ def __init__(
         )
         super().__init__(params, defaults)
 
+    # See #123400
+    @torch._disable_dynamo
```
Comment (janeyx99) [inline torch/optim/adadelta.py:47]: Looking this up in the codebase, most of the usages of torch._disable_dynamo are in optimizer code today. There are also a few in test cases, checkpointing, internal dynamo code, but is this the right way to disable dynamo? Is this the same thing as a graph break? What's the difference between this decorator and the torch patcher? Is there other precedent for disabling dynamo on library code?

Comment (mlazos) [reply torch/optim/adadelta.py:47]: This is the correct way to disable dynamo. The patcher is more indirect - having this decorator here is clearer, otherwise there is no difference. Graph breaks only avoid tracing a piece of code and resume afterward, disabling will disable dynamo on the entirety of a function and run it in python.

cc @anijain2305

Comment (anijain2305) [reply torch/optim/adadelta.py:47]: Yes, this is the right way to disable dynamo. The torch patcher is a remnant of codebase when TorchDynamo used to live outside of pytorch repo. Directly applying decorator ensures that its modified inplace, and anybody else reading from the source file will get the disabled function.

disable is different and stronger than a graph break. disable means Dynamo don't intercept on any of the frames from this function and its child functions. graph break is just a break in the function - the code before and after the break still goes through dynamo.

Comment (janeyx99) [reply torch/optim/adadelta.py:47]: Okay, even if this is the right technical approach, I am still unsure this is the direction we wanna go--why is disabling the right thing to do here? What other alternatives do we have? Is there other precedent for disabling dynamo on library code outside of optimizers?

Comment (anijain2305) [reply torch/optim/adadelta.py:47]: I think .. it depends on the priority.

If something is hard to support with torch.compile + not popular, then it brings the question - is this high priority? If not, one of the compromises is to add `disable` so that even if some rare user does use it, it does not lead to wrong results. `disable` will make it fallback to `eager`, so torch.compile will be as good as eager for this function.

We have applied `_disable_dynamo` (lazy version of `disable`) at many places in torch codebase (like don't trace inside a torch-scripted model, or don't trace inside the checkpoint API because we don't support saved_tensor_hooks).

So the question is - is this usecase very popular? I have near-zero experience with optimizers. So, I will let @janeyx99 and @mlazos decide on that.

Comment (mlazos) [reply torch/optim/adadelta.py:47]: Basically my take here is that this is loading a saved optimizer checkpoint, this will rarely occur in a model's compiled region if ever and there's zero benefit to compiling this. If we want to investigate this separately that's fine, but I think this is a valid solution to fix failures in the dynamo-traced optimizer tests and shouldn't be blocking.

Comment (janeyx99) [reply torch/optim/adadelta.py:47]: How does the nn module's equivalent `__setstate__` in module.py get handled in dynamo? How come we don't disable that in dynamo? Or is it that no test has tested this so dynamo hasn't needed to be considered for setting state for modules?

I am curious if this is a more general state API topic vs something particular to optimizers.

Comment (mlazos) [reply torch/optim/adadelta.py]: Figured out a way to remove the disable, should be good now

File: torch/testing/_internal/common_optimizers.py [hunk 3]
```diff
@@ -1747,10 +1684,10 @@ def _get_optim_inputs_including_global_cliquey_kwargs(
             ),
             DecorateInfo(
                 skipIfTorchDynamo(
-                    "fails, https://github.com/pytorch/pytorch/issues/117165"
+                    "See https://github.com/pytorch/pytorch/issues/115679"
                 ),
                 "TestOptimRenewed",
-                "test_deepcopy_copies_all_public_attrs",
+                "test_state_dict_with_cuda_params",
```
Comment (janeyx99) [inline torch/testing/_internal/common_optimizers.py:1690]: Is this change intentional? It looks diff from the rest

Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py:1690]: Yeah this is intentional, it will be able to be removed after we determine the non capturable issue.

Comment (janeyx99) [reply torch/testing/_internal/common_optimizers.py:1690]: But does this belong to this commit on the stack? i.e., does this commit cause this test_state_dict_with_cuda_params to fail now? If not, it should go elsewhere

Comment (mlazos) [reply torch/testing/_internal/common_optimizers.py]: Removed

**Review by janeyx99 — APPROVED** [review]

**Comment by mlazos** [comment]
> @pytorchbot revert -m "This broke trunk all over the place" -c ignoredsignal

How so? This just enables some tests, when I checked the hud the test_profiler tests and all the other red tests were failing before this stack. Let me know and I can help revert if needed.
